### ISSUE TITLE
failed to validate purchase - android

### ISSUE THREAD

**Issue by ddimka** [voltrue2/in-app-purchase#82]
May be someone can help with this?

{
  "signature": "FH5GkfZe3kjbKSsgf3Cwy4lQPEDKFKMyAqc0dbYCndIAdZsF9tlnmesY5NbYqlFmrScDlCE9dzO4jntnmkfmrNA9CLa8CBt8L7aLN6lfyP7/lUE2bQVZ7Tr5RIydHfCtdMNeslc7NvqpASBiBBsWD7HCZuvORZI9ZHqVUp0NlVpaaQKAyoyRHIoeoqxQdn9NljNu3CkSUVfZNzG8LdOxbdrVX5cisEjbu5IiPKccsESfihDqxjk6ha6LoPET6AsY6qgym/Cq4yUiA5/0U5AfoVI4MFw5TEViItJxVy+5VxoubuXh3Auo5xEuwDcP6OQ3GJNavBfArKn7uStbgvLYSA==",
  "data": {
    "developerPayload": "com.hippotec.maps",
    "packageName": "com.hippotec.worldifi",
    "purchaseToken": "afkknolbodjgebbdfojdpjmp.AO-J1OzkcEGlu8NSK7bRWN9OvXRK1pPE5stj7KAwQE3kqL4Nn8BhbLw4zJVL-snctdYEUB8iFdX5hgsFikniJRitbq_YA8fzZFrXPtgwQgb4MgGAbHmqJ-8CDVP96AJEvL0uDlOercRK",
    "purchaseState": 0,
    "purchaseTime": 1490616616658,
    "productId": "com.hippotec.maps"
  }
}

MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAgBcv0xfEtGxEXrC8o9/yT3Mu8+l/d1FV9WYTA1IS09i0JJeRsCrtdEf/VhAnt/mCquwKl2+d6WBi8tBYfj7n3LX2L2NDj+yhkWhs3tpDNcULIu5t0+kTyNjffzmb8wCEFJgK3/N9Zota/GVUVazQ4AyvHKB5c9vBBIHUPVds8mGMXYEhh83/AfnIIOoR5LeH0V9bwqgbi11UljcS9OYEvLb+KJzCBtRnXNLO65AoK2/eUlYI2qlOJgziJhNcEqFtRkGWFhOuk+cEYJd6twttZ3TKG9o3z8SdJauP8BJqYUwpiqf0V85+tWI4d67l53q4jvNRni7KFUwfHNdwbDQMnwIDAQAB

getting "failed to validate purchase"
Thanks

**Comment by voltrue2** [comment]
Hello,

Could you confirm that the following is actually correct?

```
MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAgBcv0xfEtGxEXrC8o9/yT3Mu8+l/d1FV9WYTA1IS09i0JJeRsCrtdEf/VhAnt/mCquwKl2+d6WBi8tBYfj7n3LX2L2NDj+yhkWhs3tpDNcULIu5t0+kTyNjffzmb8wCEFJgK3/N9Zota/GVUVazQ4AyvHKB5c9vBBIHUPVds8mGMXYEhh83/AfnIIOoR5LeH0V9bwqgbi11UljcS9OYEvLb+KJzCBtRnXNLO65AoK2/eUlYI2qlOJgziJhNcEqFtRkGWFhOuk+cEYJd6twttZ3TKG9o3z8SdJauP8BJqYUwpiqf0V85+tWI4d67l53q4jvNRni7KFUwfHNdwbDQMnwIDAQAB
```

**Comment by ddimka** [comment]
Yes. This is correct key

**Comment by pmalliak** [comment]
Any news on this? I have the same problem, is it a bug or something wrong with our keys?

**Comment by voltrue2** [comment]
Hello

I don't believe it's a bug but I need to look into this as it isn't been reported so often.

**Comment by voltrue2** [comment]
I have been running tests on this receipt against the key provided above, but it seems either the receipt or the key is or both incorrect.
How are you sending the receipt to the server to validate?
Do you have the original receipt sent from Google server?

**Comment by pmalliak** [comment]
At first I tried to validate it through the demo page. I enter the data/signature object with google key and i get "failed to validate purchase" error. Is this supposed to work?

On my project I have this on config
`  iap.config({
    googlePublicKeyPath: "../iap-sandbox"
  });`
where iap-sandbox is the a file that contains the public key,

and as receipt the data/signature object.

**Comment by voltrue2** [comment]
Yes the demo page is functional and you should be able to validate the receipt if the receipt and the key are correct.
Either of them must have been corrupt somehow. 
Would you be able to print the receipt from Google in your client app and post it here?

**Comment by pmalliak** [comment]
{ 
"data":{
"packageName":"com.programize.hirola.muchokids_android",
"productId":"test_0",
"purchaseTime":1490882619612,
"purchaseState":0,
"purchaseToken":"nppjmgcceoeiomijghpcdbof.AO-J1OyTQl1Kmv4hjCR1RWlCynlqlHn5sIsezwesBBO84RwZpiY2Ig4gPCFoeHKw35v3nWQNHR6FhZIRr_Xw-nos5v0HgKyvfdPFz54B1zwi_FlFlF2PTiWp5_rGEYRBU-H4LJnzyC1B",
"autoRenewing":true
}, "signature":"GJvrduu0rIll1CCa2EwfBobNx58PUcH9n70IK6WNDuM5IQX/MhNtt5Z2ywD0pTZlynQcxIXO1N8isIhwQuq0+Z3wk5iMXAmXfbdIjFYtZVTp4zS+EJpPAWpt1n5cAG67MiOqKQ0ZRXvXxZGXHpL5a0ws6UsG7EtPA4SaVozuGdFMfSrqNybpjeLb9mkZ2b0HF5zP+PhgtS78CUAxH78N4dT+w9fm0SOWWVlZiePPEL7ivy0jcoa7Av1htHzRies65iAiByw2q0sbVxm0XeWzQPKioEGTV8FdGp0S5tuIiUzhoW1cdNXjwnQ+g2sV1JR5mjQ3nwVpbCUdpxUHNa7O7g==" 
}

The data object is the receipt from google. And the whole object is what I m trying to validate on demo page. 

ps. This is a subscription on a test environment

**Comment by voltrue2** [comment]
the receipt seems to be an auto-renewable subscription.
Have you followed these steps to properly set up the subscription type items?

https://github.com/voltrue2/in-app-purchase#google-play-store-api-1

**Comment by ddimka** [comment]
Hi. Just now tryed to buy an item.
got: 

purchaseData
{"orderId":"GPA.3398-1913-5131-24467","packageName":"com.hippotec.worldifi","productId":"com.hippotec.maps","purchaseTime":1491390709830,"purchaseState":0,"developerPayload":"com.hippotec.maps","purchaseToken":"niddidlgnjomfnplmegdhail.AO-J1OxC3wZs3ofI_wViPmzJ_bdLP-gW0wt8S-O2LTzVkgfV0yAzK949oBoI5fglzSpUUtvdVuPU7m1gXLB8SuVsMlhGwJpMAdAXgYMjfeNG4FMmTGRyrYxHfbtOocLaqnDMdXOQRlQj"}


Signature:
Qwb2yE0nhYJhxatIrjKGXFkOwhSLKXkSHpMgsrDjXE6S9LFcFtxq1ECACmDYWLljsNFqhE6pdxOhc0NjmYyg5bJ3C8bGK1/QvlZmjH3nyYageUdPhq3tArBfZQPs+Mv9CkTIKYnqtFAC0sKros+mSCZbX+lNGO/QKTxWmyRiQtYeF2PpgA2iANW9Llt1uM7X/k5NgSo48qfF/RT5asCIDFMKkGIZ0DSXfMsCD9ZNC+N5crrhx3VaBzGUBgZgjWytHI1assA7e39DI8SdWPPFCIJ2HX3e2q7d5nrEjgKkFGGQF9CCgFMOBpT+ab490+1vGLfjJrbXCS/7vF9wpFFYFA==

Still error with validation.

**Comment by ddimka** [comment]
May be this can help..
[1491393720623][VERBOSE] <Google> Validate this: { signature: 'Qwb2yE0nhYJhxatIrjKGXFkOwhSLKXkSHpMgsrDjXE6S9LFcFtxq1ECACmDYWLljsNFqhE6pdxOhc0NjmYyg5bJ3C8bGK1/QvlZmjH3nyYageUdPhq3tArBfZQPs+Mv9CkTIKYnqtFAC0sKros+mSCZbX+lNGO/QKTxWmyRiQtYeF2PpgA2iANW9Llt1uM7X/k5NgSo48qfF/RT5asCIDFMKkGIZ0DSXfMsCD9ZNC+N5crrhx3VaBzGUBgZgjWytHI1assA7e39DI8SdWPPFCIJ2HX3e2q7d5nrEjgKkFGGQF9CCgFMOBpT+ab490+1vGLfjJrbXCS/7vF9wpFFYFA==',
  data: 
   { developerPayload: 'com.hippotec.maps',
     packageName: 'com.hippotec.worldifi',
     purchaseToken: 'niddidlgnjomfnplmegdhail.AO-J1OxC3wZs3ofI_wViPmzJ_bdLP-gW0wt8S-O2LTzVkgfV0yAzK949oBoI5fglzSpUUtvdVuPU7m1gXLB8SuVsMlhGwJpMAdAXgYMjfeNG4FMmTGRyrYxHfbtOocLaqnDMdXOQRlQj',
     purchaseState: 0,
     orderId: 'GPA.3398-1913-5131-24467',
     purchaseTime: 1491390709830,
     productId: 'com.hippotec.maps' } }
[1491393720631][VERBOSE] <Google> Auto stringified receipt data: {"developerPayload":"com.hippotec.maps","packageName":"com.hippotec.worldifi","purchaseToken":"niddidlgnjomfnplmegdhail.AO-J1OxC3wZs3ofI_wViPmzJ_bdLP-gW0wt8S-O2LTzVkgfV0yAzK949oBoI5fglzSpUUtvdVuPU7m1gXLB8SuVsMlhGwJpMAdAXgYMjfeNG4FMmTGRyrYxHfbtOocLaqnDMdXOQRlQj","purchaseState":0,"orderId":"GPA.3398-1913-5131-24467","purchaseTime":1491390709830,"productId":"com.hippotec.maps"}
[1491393720631][VERBOSE] <Google> Try validate against live public key: MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAgBcv0xfEtGxEXrC8o9/yT3Mu8+l/d1FV9WYTA1IS09i0JJeRsCrtdEf/VhAnt/mCquwKl2+d6WBi8tBYfj7n3LX2L2NDj+yhkWhs3tpDNcULIu5t0+kTyNjffzmb8wCEFJgK3/N9Zota/GVUVazQ4AyvHKB5c9vBBIHUPVds8mGMXYEhh83/AfnIIOoR5LeH0V9bwqgbi11UljcS9OYEvLb+KJzCBtRnXNLO65AoK2/eUlYI2qlOJgziJhNcEqFtRkGWFhOuk+cEYJd6twttZ3TKG9o3z8SdJauP8BJqYUwpiqf0V85+tWI4d67l53q4jvNRni7KFUwfHNdwbDQMnwIDAQAB
[1491393720633][VERBOSE] <Google> Failed to validate against: MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAgBcv0xfEtGxEXrC8o9/yT3Mu8+l/d1FV9WYTA1IS09i0JJeRsCrtdEf/VhAnt/mCquwKl2+d6WBi8tBYfj7n3LX2L2NDj+yhkWhs3tpDNcULIu5t0+kTyNjffzmb8wCEFJgK3/N9Zota/GVUVazQ4AyvHKB5c9vBBIHUPVds8mGMXYEhh83/AfnIIOoR5LeH0V9bwqgbi11UljcS9OYEvLb+KJzCBtRnXNLO65AoK2/eUlYI2qlOJgziJhNcEqFtRkGWFhOuk+cEYJd6twttZ3TKG9o3z8SdJauP8BJqYUwpiqf0V85+tWI4d67l53q4jvNRni7KFUwfHNdwbDQMnwIDAQAB Error: failed to validate purchase
    at validatePublicKey (D:\Projects\WebstormProjects\wonderfi_server\node_modules\in-app-purchase\lib\google.js:409:5)
    at Object.module.exports.validatePurchase (D:\Projects\WebstormProjects\wonderfi_server\node_modules\in-app-purchase\lib\google.js:161:2)
    at Object.module.exports.validate (D:\Projects\WebstormProjects\wonderfi_server\node_modules\in-app-purchase\index.js:45:11)
    at D:\Projects\WebstormProjects\wonderfi_server\routes\purchase.js:274:31
    at D:\Projects\WebstormProjects\wonderfi_server\node_modules\in-app-purchase\lib\async.js:60:5
    at Object.module.exports.setup (D:\Projects\WebstormProjects\wonderfi_server\node_modules\in-app-purchase\lib\amazon.js:50:10)
    at D:\Projects\WebstormProjects\wonderfi_server\node_modules\in-app-purchase\index.js:34:11
    at iterator (D:\Projects\WebstormProjects\wonderfi_server\node_modules\in-app-purchase\lib\async.js:54:3)
    at Object.module.exports.parallel (D:\Projects\WebstormProjects\wonderfi_server\node_modules\in-app-purchase\lib\async.js:65:3)
    at Object.module.exports.setup (D:\Projects\WebstormProjects\wonderfi_server\node_modules\in-app-purchase\index.js:26:8)
    at validatePurchase (D:\Projects\WebstormProjects\wonderfi_server\routes\purchase.js:261:27)
    at D:\Projects\WebstormProjects\wonderfi_server\routes\purchase.js:115:21
    at D:\Projects\WebstormProjects\wonderfi_server\routes\purchase.js:310:17
    at D:\Projects\WebstormProjects\wonderfi_server\node_modules\keystone\node_modules\mongoose\lib\query.js:2257:18
    at tryCatch (D:\Projects\WebstormProjects\wonderfi_server\node_modules\keystone\node_modules\es6-promise\dist\es6-promise.js:410:12)
    at invokeCallback (D:\Projects\WebstormProjects\wonderfi_server\node_modules\keystone\node_modules\es6-promise\dist\es6-promise.js:425:13)
    at publish (D:\Projects\WebstormProjects\wonderfi_server\node_modules\keystone\node_modules\es6-promise\dist\es6-promise.js:393:7)
    at flush (D:\Projects\WebstormProjects\wonderfi_server\node_modules\keystone\node_modules\es6-promise\dist\es6-promise.js:121:5)
    at _combinedTickCallback (internal/process/next_tick.js:67:7)
    at process._tickCallback (internal/process/next_tick.js:98:9)

**Comment by voltrue2** [comment]
This issue seems to be also related to https://github.com/voltrue2/in-app-purchase/issues/113
Please have a look. #113 has been solved with `1.6.0` and later version of the module.

**Comment by voltrue2** [comment]
Closing as there has not been a response for a while.
Please reopen this issue if there still is an issue.
